src/lib/datocms/index.ts

```diff
@@ -16,12 +17,50 @@ type DatocmsRequest = {
   variables?: { [key: string]: string };
   retryCount?: number;
 };
+
+/**
+ * Expect return value of specified Query to be available, i.e. non-nullable.
+ */
+type Assumed<Query> = Required<{ [key in keyof Query]: NonNullable<Query[key]> }>;
+
 /**
  * Makes a request to the DatoCMS GraphQL API using the provided query and variables.
  * It has authorization, environment and drafts (preview) pre-configured.
  * It has a retry mechanism in case of rate-limiting, based on DatoCMS API utils. @see https://github.com/datocms/js-rest-api-clients/blob/f4e820d/packages/rest-client-utils/src/request.ts#L239C13-L255
+ * 
+ * @template Query - The expected response data structure
+ * 
+ * @param options - Request options
+ * @param options.query - The GraphQL query document
+ * @param options.variables - Variables to pass to the GraphQL query
+ * @param options.retryCount - Number of retry attempts in case of rate limiting
+ * @param options.prerender - When true, forces non-nullable values in the response, (default is based on Astro output mode)
+ * 
+ * @returns A promise resolving to the query result, with forced non-null values when prerender is true
  */
-export const datocmsRequest = async <T>({ query, variables = {}, retryCount = 1 }: DatocmsRequest): Promise<T> => {
+export async function datocmsRequest<Query>(
+  options: DatocmsRequest & { prerender: true }
+): Promise<Assumed<Query>>;
+export async function datocmsRequest<Query>(
+  options: DatocmsRequest & { prerender: false }
+): Promise<Query>;
+export async function datocmsRequest<
+  Query,
+  Prerender extends boolean = typeof output extends 'static' ? true : false
+>(
+  options: DatocmsRequest & { prerender?: Prerender }
+): Promise<Prerender extends true ? Assumed<Query> : Query>;
+export async function datocmsRequest<
+  Query,
+  Prerender extends boolean = typeof output extends 'static' ? true : false
+>(
+  {
+    query,
+    variables = {},
+    retryCount = 1,
+    prerender = (output as OutputModes === 'static') as Prerender,
```

Comment:
By widening the type of `output` with `as OutputModes`, the typing of this line is still correct, even when `output` would be set to `server`. All other definitions are still inferred from the value of `output` itself.